lib/url.js

```diff
@@ -10,60 +10,73 @@ exports.format = urlFormat;
 exports.Url = Url;
 
 function Url() {
-  this.protocol = null;
-  this.slashes = null;
+  // For more efficient internal representation and laziness.
+  // The non-underscore versions of these properties are accessor functions
+  // defined on the prototype.
+  this._protocol = null;
+  this._href = '';
+  this._port = -1;
+  this._query = null;
+  this._prependSlash = false;
   this.auth = null;
+  this.slashes = null;
   this.host = null;
-  this.port = null;
   this.hostname = null;
   this.hash = null;
   this.search = null;
-  this.query = null;
   this.pathname = null;
-  this.path = null;
-  this.href = null;
 }
 
 // Reference: RFC 3986, RFC 1808, RFC 2396
```

Comment:
At some point we may want to update this to say that our canonical source of info is https://url.spec.whatwg.org/ which obsoletes those RFCs. But we are not compliant with any standard yet, really.